typing/typecore.ml

```diff
@@ -670,92 +674,109 @@ end) = struct
         in
         List.find check_type lbls
 
+  (** [disambiguate] selects a concrete description for [lid] using some
+      contextual information: an optional [expected_type], and a list of candidates [lbls].
+      If [expected_type] is [None], it just returns the head of [lbls].
+      Otherwise, it extracts the description directly from the corresponding type definition,
+      except for extension types, where it has to choose it inside [lbls].
+      This code is careful to warn in a number of situations:
+      * if a non principal type was used for disambiguation [Not_principal]
+      * if [lbls] contained several valid candidates [Ambiguous_name]
```

Comment:
Why would that matter if type-based selection is used?

Reply:
This matters because, when the type is not known principally, resolution without the type could have given a different result.